src/librustc_mir/interpret/memory.rs

```diff
@@ -791,6 +791,30 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> Memory<'mir, 'tcx, M> {
         self.get_raw(ptr.alloc_id)?.read_c_str(self, ptr)
     }
 
+    /// Reads a 0x0000-terminated u16-sequence from memory. Returns them as a Vec<u16>.
+    /// Terminator 0x0000 is not included in the returned Vec<u16>.
+    ///
+    /// Performs appropriate bounds checks.
+    pub fn read_wide_str(&self, ptr: Scalar<M::PointerTag>) -> InterpResult<'tcx, Vec<u16>> {
+        let size_2bytes = Size::from_bytes(2);
+        let align_2bytes = Align::from_bytes(2).unwrap();
+        let mut ptr = self.check_ptr_access(ptr, size_2bytes, align_2bytes)?.unwrap();
+        let allocation = self.get_raw(ptr.alloc_id)?;
+        let mut u16_seq = Vec::new();
+
+        loop {
+            let single_u16 = allocation.read_scalar(self, ptr, size_2bytes)?.to_u16()?;
+            if single_u16 != 0x0000 {
+                u16_seq.push(single_u16);
+                let scalar = std::convert::From::from(ptr.offset(size_2bytes, self)?);
+                ptr = self.check_ptr_access(scalar, size_2bytes, align_2bytes)?.unwrap();
```

Comment:
Instead of duplicating the bounds check, it is probably better to do it just directly before `read_scalar`. That also makes it more obvious that indeed every `read_scalar` is appropriately guarded.